helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -697,10 +706,16 @@ void createClient() throws Exception {
     int retryCount = 0;
     while (retryCount < 3) {
       try {
-        // TODO: synchronize this block and wait for the new non-zero session ID updated.
-        _zkclient.waitUntilConnected(_connectionInitTimeout, TimeUnit.MILLISECONDS);
-        handleStateChanged(KeeperState.SyncConnected);
-        handleNewSession();
+        // TODO: might be better to create a new method to return session id in waitUntilConnected.
+        synchronized (this) {
```

Comment:
Also be caution waitUntilConnected() is a blocking call, make sure the lock is released within that method if it goes to sleep.